cvmfs/cache_ram.cc

```diff
@@ -96,38 +93,38 @@ int64_t RamCacheManager::GetSize(int fd) {
     LogCvmfs(kLogCache, kLogDebug, "bad fd %d on GetSize", fd);
     return -EBADF;
   }
-  assert(open_fds_[fd].store);
   perf::Inc(counters_.n_getsize);
-  return open_fds_[fd].store->GetSize(open_fds_[fd].handle);
+  return GetStore(open_fds_[fd])->GetSize(open_fds_[fd].handle);
 }
 
 
 int RamCacheManager::Close(int fd) {
   bool rc;
-  bool sweep_tail = false;
 
   WriteLockGuard guard(rwlock_);
   if (!IsValid(fd)) {
     LogCvmfs(kLogCache, kLogDebug, "bad fd %d on Close", fd);
     return -EBADF;
   }
-  assert(open_fds_[fd].store);
-  rc = open_fds_[fd].store->Unref(open_fds_[fd].handle);
+  rc = GetStore(open_fds_[fd])->Unref(open_fds_[fd].handle);
   assert(rc);
   open_fds_[fd].handle = kInvalidHandle;
-  LogCvmfs(kLogCache, kLogDebug, "closed fd %d", fd);
-  sweep_tail = (static_cast<unsigned>(fd) == (open_fds_.size() - 1));
-
-  if (sweep_tail) {
-    unsigned last_good_idx = open_fds_.size() - 1;
-    while (last_good_idx > 0 &&
-           open_fds_[last_good_idx].handle == kInvalidHandle)
-      last_good_idx--;
-    open_fds_.resize(last_good_idx + 1);
-    LogCvmfs(kLogCache, kLogDebug, "resized fd vector to %u",
-             last_good_idx + 1);
-  }
 
+  size_t index = open_fds_[fd].index;
+  assert(index < fd_index_.size());
+  assert(fd_pivot_ < fd_index_.size());
```

Comment:
I'm using `fd_pivot_` to refer to the next free spot, so it's initialized to zero, and if it reaches `fd_index_.size()`, we've run off the end of the array. To be sure I'm not making any off-by-one errors, a unit test to hammer the fds sounds like a good idea